File: Assets/Scripts/CSharp/Tank/TankShooting.cs
using UnityEngine;
using UnityEngine.UI;
using System;
using System.Collections.Generic;

namespace PuertsTest
{
    public class TankShooting : WeChat.PuertsBeefBallBehaviour
    {
        public int m_PlayerNumber = 1;              // Used to identify the different players.
        //public Rigidbody m_Shell;                   // Prefab of the shell.
        // public GameObject m_Shell;                   // Prefab of the shell.

        public string m_ShellPath;
        public Transform m_FireTransform;           // A child of the tank where the shells are spawned.
        public Slider m_AimSlider;                  // A child of the tank that displays the current launch force.
        // public AudioSource m_ShootingAudio;         // Reference to the audio source used to play the shooting audio. NB: different to the movement audio source.
        // public AudioClip m_ChargingClip;            // Audio that plays when each shot is charging up.
        // public AudioClip m_FireClip;                // Audio that plays when each shot is fired.
        public float m_MinLaunchForce = 15f;        // The force given to the shell if the fire button is not held.
        public float m_MaxLaunchForce = 30f;        // The force given to the shell if the fire button is held for the max charge time.
        public float m_MaxChargeTime = 0.75f;       // How long the shell can charge for before it is fired at max force.
        public float m_FireInterval;
        public PhyWorld m_PhyWorld = null;


        new string JSClassName = "src/Tank/TankShooting";
    }
}

File: Typescript/src/Tank/TankShooting.ts
import ShellExplosion from "../Shell/ShellExplosion";
import ShellPool from "../Shell/ShellPool";

export default class TankShooting extends PuertsLogic<CS.PuertsTest.TankShooting> {

  private m_FireButton: string;                // The input axis that is used for launching shells.
  private m_CurrentLaunchForce: number;         // The force that will be given to the shell when the fire button is released.
  private m_ChargeSpeed: number;                // How fast the launch force increases, based on the max charge time.
  //private bool m_Fired;                       // Whether or not the shell has been launched with this button press.
  private m_Shell: PuertsBeefBallSDK.Prefab = null;


  lastT: Date = new Date;
  private m_ShellInstanceList: Array<ShellExplosion> = [];


  OnEnable(): void {
    // When the tank is turned on, reset the launch force and the UI
    this.m_CurrentLaunchForce = this.behaviour.m_MinLaunchForce;
    //m_AimSlider.value = m_MinLaunchForce;
    PuertsBeefBallSDK.load<PuertsBeefBallSDK.Prefab>(this.behaviour.m_ShellPath)
      .then((prefab: PuertsBeefBallSDK.Prefab) => {
        CS.UnityEngine.Debug.Log(!!prefab);
        this.m_Shell = prefab;
      })
  }


  Start(): void {
    // The fire axis is based on the player number.
    this.m_FireButton = "Fire" + this.behaviour.m_PlayerNumber;

    // The rate that the launch force charges up is the range of possible forces by the max charge time.
    this.m_ChargeSpeed = (this.behaviour.m_MaxLaunchForce - this.behaviour.m_MinLaunchForce) / this.behaviour.m_MaxChargeTime;
  }


  public Update() {
    var time = new Date;
    var dt = (time.getTime() - this.lastT.getTime()) / 1000;
    if (dt >= this.behaviour.m_FireInterval && this.m_Shell != null) {
      this.Fire();
      this.lastT = time;
    }
  }


  private Fire(): void {
    //// Set the fired flag so only Fire is only called once.
    //m_Fired = true;

    var shellInstance: CS.UnityEngine.GameObject = ShellPool.Create(
      this.m_Shell,
      this.behaviour.m_FireTransform.transform.position,
      this.behaviour.m_FireTransform.transform.rotation
    ) as CS.UnityEngine.GameObject;
    var shell: CS.PuertsTest.ShellExplosion = shellInstance.GetComponent(puerts.$typeof(CS.PuertsTest.ShellExplosion)) as CS.PuertsTest.ShellExplosion;
    shell.Reset();
    shell.m_PhyWorld = (this.behaviour as any).m_PhyWorld;
    shell.m_PlayerNumber = (this.behaviour as any).m_PlayerNumber;

  }
}